src/physics/em/KNInteractor.i.hh

```diff
@@ -0,0 +1,130 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file KNInteractor.i.hh
+//---------------------------------------------------------------------------//
+
+#include "base/ArrayUtils.hh"
+#include "base/Constants.hh"
+#include "random/distributions/BernoulliDistribution.hh"
+#include "random/distributions/GenerateCanonical.hh"
+#include "random/distributions/UniformRealDistribution.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ *
+ * The incident particle must be above the energy threshold: this should be
+ * handled in code *before* the interactor is constructed.
+ */
+KNInteractor::KNInteractor(const KNInteractorPointers& shared,
+                           const ParticleTrackView&    particle,
+                           const Real3&                inc_direction,
+                           SecondaryAllocatorView&     allocate)
+    : shared_(shared)
+    , inc_energy_(particle.energy())
+    , inc_direction_(inc_direction)
+    , allocate_(allocate)
+{
+    REQUIRE(inc_energy_ >= this->min_incident_energy());
+    REQUIRE(particle.def_id() == shared_.gamma_id);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample Compton scattering using the Klein-Nishina model.
+ *
+ * See section 6.4.2 of the Geant physics reference. Epsilon is the ratio of
+ * outgoing to incident gamma energy, bounded in [epsilon_0, 1].
+ */
+template<class Engine>
+CELER_FUNCTION Interaction KNInteractor::operator()(Engine& rng)
```

Comment:
This seems "large" of an inline function.  What is our recommended threshold for a function to be in the .i.hh vs the .cc?

Reply:
Since I'd like to avoid CUDA separable compilation, all of the host/device helper functions like this one will have to be in header files, since they'll be included by `Foo.test.cc` as well as `KernelThatUsesFoo.cu`. Because these functions will only be used in tests and kernel implementations, they won't need to be included in many places. (Most of the physics interactors I expect will only be included by two translation units. Others, like ParticleTrackView, will be included in quite a few.)

Reply:
But you do make a good point, if this were a non-templated host-only function it should definitely be in a `.cc` file.

Reply:
I did indeed miss that this was templated.  

> Since I'd like to avoid CUDA separable compilation

As a side note, this is probably avoidable only in the short term.